File: backend/app/canvas/text_measure.py
# === FILE: backend/app/canvas/text_measure.py ===

"""
Server-side text measurement for Excalidraw.

IMPORTANT: This is an APPROXIMATION used for initial placement only.
The frontend uses @chenglou/pretext for pixel-perfect measurements
and corrects positions via sync.

Calibrated against Pretext output for Excalidraw's default fonts.
"""

from __future__ import annotations
import math

# ── Character width tables at fontSize=16 ──
# Calibrated against Pretext measurements for Excalidraw fonts.
# Each value = average advance width per character in pixels at 16px.
_BASE_WIDTHS = {
    1: 9.2,    # Virgil (handwritten)
    2: 8.4,    # Helvetica
    3: 9.6,    # Cascadia (monospace — very predictable)
    4: 8.8,    # Excalifont
    5: 8.2,    # Nunito
    6: 9.0,    # Lilita One
    7: 8.6,    # Comic Shanns
}

# Width multipliers for character classes
_NARROW = set("iljtfr!|[](){}.,:;'\"1")
_WIDE = set("mwMWOQGD@#%&")


def _char_width(ch: str, base: float) -> float:
    if ch in _NARROW:
        return base * 0.5
    if ch in _WIDE:
        return base * 1.35
    if ch == " ":
        return base * 0.35
    if ch == "\t":
        return base * 1.6
    if ch.isupper():
        return base * 1.1
    return base


def _line_width(text: str, base: float) -> float:
    if not text:
        return 0.0
    return sum(_char_width(ch, base) for ch in text)


def measure_text(
    text: str,
    font_size: int = 16,
    font_family: int = 1,
    max_width: float = 600,
    max_lines: int = 200,
) -> dict:
    """
    Approximate text wrapping and measurement for Excalidraw.
    Returns: {wrapped_text, width, height, line_count}
    """
    if not text or not text.strip():
        lh = font_size * 1.25
        return {"wrapped_text": "", "width": 20, "height": lh + 4, "line_count": 1}

    scale = font_size / 16.0
    base = _BASE_WIDTHS.get(font_family, 9.2) * scale
    line_height = font_size * 1.25

    # Excalidraw's actual rendering has some internal padding
    effective_max = max_width * 0.95

    paragraphs = text.split("\n")
    all_lines: list[str] = []

    for para in paragraphs:
        if len(all_lines) >= max_lines:
            break

        if not para.strip():
            all_lines.append("")
            continue

        words = para.split(" ")
        current_line = ""
        current_w = 0.0

        for word in words:
            word_w = _line_width(word, base)

            # Force-break overly long words
            if word_w > effective_max and not current_line:
                chunk = ""
                chunk_w = 0.0
                for ch in word:
                    cw = _char_width(ch, base)
                    if chunk_w + cw > effective_max and chunk:
                        all_lines.append(chunk)
                        if len(all_lines) >= max_lines:
                            break
                        chunk = ch
                        chunk_w = cw
                    else:
                        chunk += ch
                        chunk_w += cw
                if chunk and len(all_lines) < max_lines:
                    current_line = chunk
                    current_w = chunk_w
                continue

            sep = " " if current_line else ""
            sep_w = _line_width(sep, base)

            if current_w + sep_w + word_w <= effective_max or not current_line:
                current_line += sep + word
                current_w += sep_w + word_w
            else:
                all_lines.append(current_line)
                if len(all_lines) >= max_lines:
                    break
                current_line = word
                current_w = word_w

        if current_line and len(all_lines) < max_lines:
            all_lines.append(current_line)

    if len(all_lines) > max_lines:
        all_lines = all_lines[:max_lines]
        if all_lines:
            all_lines[-1] = all_lines[-1].rstrip() + "…"

    wrapped = "\n".join(all_lines)
    actual_w = max((_line_width(l, base) for l in all_lines), default=20)
    actual_h = len(all_lines) * line_height + 4

    return {
        "wrapped_text": wrapped,
        "width": min(max(actual_w, 20), max_width),
        "height": max(actual_h, line_height + 4),
        "line_count": len(all_lines),
    }


def measure_text_batch(requests: list[dict]) -> list[dict]:
    return [
        measure_text(
            text=r.get("text", ""),
            font_size=r.get("font_size", 16),
            font_family=r.get("font_family", 1),
            max_width=r.get("max_width", 600),
            max_lines=r.get("max_lines", 200),
        )
        for r in requests
    ]

File: frontend/src/lib/textMeasure.ts
// === FILE: frontend/src/lib/textMeasure.ts ===

/**
 * Text measurement using @chenglou/pretext.
 * Single source of truth for text dimensions across the entire app.
 * 
 * Why: The backend's character-width lookup tables were ~20-30% off,
 * causing text overflow, overlapping cards, and broken diagram labels.
 * Pretext uses the browser's actual Canvas font engine → pixel-perfect.
 */

import { prepare, layout } from "@chenglou/pretext"

// ── Excalidraw font families → CSS font strings ──
const FONT_MAP: Record<number, string> = {
  1: '"Virgil"',
  2: '"Helvetica"',
  3: '"Cascadia"',
  4: '"Excalifont"',
  5: '"Nunito"',
  6: '"Lilita One"',
  7: '"Comic Shanns"',
}

// Cache prepared text handles (Pretext recommends reusing these)
const _cache = new Map<string, ReturnType<typeof prepare>>()

function getCacheKey(text: string, font: string): string {
  return `${font}::${text.slice(0, 200)}`
}

export interface TextMeasurement {
  width: number
  height: number
  lineCount: number
  wrappedText: string
}

/**
 * Measure text exactly as Excalidraw will render it.
 * Uses Pretext for pixel-perfect accuracy.
 */
export function measureText(
  text: string,
  fontSize: number = 16,
  fontFamily: number = 1,
  maxWidth: number = 600,
  maxLines: number = 200,
): TextMeasurement {
  if (!text || !text.trim()) {
    const lineHeight = fontSize * 1.25
    return { width: 20, height: lineHeight + 4, lineCount: 1, wrappedText: "" }
  }

  const fontName = FONT_MAP[fontFamily] || FONT_MAP[1]
  const fontStr = `${fontSize}px ${fontName}`
  const lineHeight = fontSize * 1.25

  const key = getCacheKey(text, fontStr)
  let prepared = _cache.get(key)
  if (!prepared) {
    try {
      prepared = prepare(text, fontStr)
      // Keep cache bounded
      if (_cache.size > 500) {
        const firstKey = _cache.keys().next().value
        if (firstKey) _cache.delete(firstKey)
      }
      _cache.set(key, prepared)
    } catch {
      // Fallback if font not loaded yet
      return fallbackMeasure(text, fontSize, maxWidth, maxLines)
    }
  }

  const result = layout(prepared, maxWidth, lineHeight)

  // Clamp to max lines
  let finalHeight = result.height
  let lineCount = result.lineCount
  if (lineCount > maxLines) {
    lineCount = maxLines
    finalHeight = maxLines * lineHeight
  }

  return {
    width: Math.min(Math.max(finalHeight > lineHeight * 1.5 ? maxWidth : maxWidth * 0.8, 20), maxWidth),
    height: Math.max(finalHeight + 4, lineHeight + 4),
    lineCount,
    wrappedText: text, // Pretext doesn't wrap — Excalidraw handles wrapping
  }
}

/**
 * Measure a batch of texts (for placement calculations).
 */
export function measureBatch(
  items: Array<{
    text: string
    fontSize?: number
    fontFamily?: number
    maxWidth?: number
  }>,
): TextMeasurement[] {
  return items.map((item) =>
    measureText(
      item.text,
      item.fontSize ?? 16,
      item.fontFamily ?? 1,
      item.maxWidth ?? 600,
    ),
  )
}

/**
 * Compute the actual rendered height of a note card.
 * Used by the placement engine to avoid overlaps.
 */
export function measureNoteCard(
  title: string,
  summary: string,
  tags: string[],
  cardWidth: number = 360,
): { totalHeight: number; titleHeight: number; summaryHeight: number } {
  const contentWidth = cardWidth - 24 // 12px padding each side

  const titleM = measureText(title || "Untitled", 18, 1, contentWidth, 2)
  const summaryM = measureText(summary || "", 13, 1, contentWidth, 6)
  const tagHeight = tags.length > 0 ? 20 : 0

  const totalHeight = 24 + titleM.height + 8 + summaryM.height + tagHeight + 24

  return {
    totalHeight: Math.max(totalHeight, 120), // minimum card height
    titleHeight: titleM.height,
    summaryHeight: summaryM.height,
  }
}

/**
 * Compute diagram node dimensions from label text.
 */
export function measureDiagramNode(
  label: string,
  requestedWidth: number = 180,
  maxWidth: number = 300,
): { width: number; height: number } {
  const innerWidth = Math.min(requestedWidth, maxWidth) - 24
  const m = measureText(label, 14, 1, innerWidth, 3)

  return {
    width: Math.min(Math.max(requestedWidth, m.width + 32), maxWidth),
    height: Math.max(56, m.height + 24),
  }
}

// ── Fallback for when fonts aren't loaded yet ──
function fallbackMeasure(
  text: string,
  fontSize: number,
  maxWidth: number,
  maxLines: number,
): TextMeasurement {
  const avgCharWidth = fontSize * 0.6
  const lineHeight = fontSize * 1.25
  const charsPerLine = Math.floor(maxWidth / avgCharWidth)
  const lines = Math.min(Math.ceil(text.length / charsPerLine), maxLines)

  return {
    width: Math.min(text.length * avgCharWidth, maxWidth),
    height: lines * lineHeight + 4,
    lineCount: lines,
    wrappedText: text,
  }
}

/**
 * Clear the measurement cache (call on font changes).
 */
export function clearMeasurementCache() {
  _cache.clear()
}